// angular-geocoder/__ivy_ngcc__/fesm2015/angular-geocoder.js
import { Injectable, EventEmitter, Component, ViewChild, Input, Output, HostListener, NgModule } from '@angular/core';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { CommonModule } from '@angular/common';
import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
import { map, filter, debounceTime, distinctUntilChanged, switchMap } from 'rxjs/operators';
import { parse } from 'terraformer-wkt-parser';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common/http';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '@angular/common';

const _c0 = ["geocoder"];
function GeocoderComponent_div_11_div_1_li_2_p_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "p", 23);
} if (rf & 2) {
    const place_r6 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", place_r6.highlight, ɵngcc0.ɵɵsanitizeHtml);
} }
function GeocoderComponent_div_11_div_1_li_2_p_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const field_r12 = ctx.$implicit;
    const place_r6 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", place_r6[field_r12], " ");
} }
function GeocoderComponent_div_11_div_1_li_2_p_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 24);
    ɵngcc0.ɵɵtemplate(1, GeocoderComponent_div_11_div_1_li_2_p_2_ng_container_1_Template, 2, 1, "ng-container", 25);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r9.resultFieldsToShow);
} }
const _c1 = function (a0) { return { "focus-background": a0 }; };
function GeocoderComponent_div_11_div_1_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 19);
    ɵngcc0.ɵɵlistener("click", function GeocoderComponent_div_11_div_1_li_2_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r15); const place_r6 = ctx.$implicit; const ctx_r14 = ɵngcc0.ɵɵnextContext(3); return ctx_r14.lookup(place_r6.id); });
    ɵngcc0.ɵɵtemplate(1, GeocoderComponent_div_11_div_1_li_2_p_1_Template, 1, 1, "p", 20);
    ɵngcc0.ɵɵtemplate(2, GeocoderComponent_div_11_div_1_li_2_p_2_Template, 2, 1, "p", 21);
    ɵngcc0.ɵɵelementStart(3, "p", 22);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const place_r6 = ctx.$implicit;
    const i_r7 = ctx.index;
    const ctx_r5 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵpropertyInterpolate("tabindex", i_r7);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx_r5.isHighlighted(i_r7)));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.resultFieldsToShow.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.resultFieldsToShow.length > 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", place_r6.type, " ");
} }
function GeocoderComponent_div_11_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "ul", 17);
    ɵngcc0.ɵɵtemplate(2, GeocoderComponent_div_11_div_1_li_2_Template, 5, 7, "li", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.places);
} }
function GeocoderComponent_div_11_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelementStart(1, "p");
    ɵngcc0.ɵɵtext(2, "Geen zoekresultaten gevonden voor ");
    ɵngcc0.ɵɵelementStart(3, "strong");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(5, ".");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r3.searchField.value, "");
} }
function GeocoderComponent_div_11_div_3_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 30);
    ɵngcc0.ɵɵlistener("click", function GeocoderComponent_div_11_div_3_li_2_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r20); const collation_r17 = ctx.$implicit; const ctx_r19 = ɵngcc0.ɵɵnextContext(3); return ctx_r19.onCollationClick(collation_r17.naam); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const collation_r17 = ctx.$implicit;
    const i_r18 = ctx.index;
    ɵngcc0.ɵɵpropertyInterpolate("tabindex", i_r18);
    ɵngcc0.ɵɵproperty("innerHTML", collation_r17.weergavenaam, ɵngcc0.ɵɵsanitizeHtml);
} }
function GeocoderComponent_div_11_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵelementStart(1, "ul", 28);
    ɵngcc0.ɵɵtemplate(2, GeocoderComponent_div_11_div_3_li_2_Template, 1, 2, "li", 29);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r4.collations);
} }
function GeocoderComponent_div_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵnamespaceHTML();
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtemplate(1, GeocoderComponent_div_11_div_1_Template, 3, 1, "div", 13);
    ɵngcc0.ɵɵtemplate(2, GeocoderComponent_div_11_div_2_Template, 6, 1, "div", 14);
    ɵngcc0.ɵɵtemplate(3, GeocoderComponent_div_11_div_3_Template, 3, 1, "div", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.places.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.isNoResultsFound());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showCollations());
} }
const _c2 = function (a0) { return { "highlight": a0 }; };
class GeocoderService {
    constructor(http) {
        this.http = http;
        this.geocoderBaseUrl = 'https://geodata.nationaalgeoregister.nl/locatieserver/v3';
    }
    suggest(query, options) {
        let params = {
            q: query,
            fq: '*',
            start: 0,
            rows: 10,
        };
        if (options) {
            params = Object.assign(params, options);
        }
        return this.http.get(`${this.geocoderBaseUrl}/suggest?`, { params: params })
            .toPromise().then((suggestResultObject) => {
            const collations = this.formatCollations(suggestResultObject.spellcheck.collations);
            const places = this.formatPlaces(suggestResultObject);
            return { collations, places };
        });
    }
    suggest$(query, options) {
        let params = {
            q: query,
            fq: '*',
            start: 0,
        };
        if (options) {
            params = Object.assign(params, options);
        }
        return this.http.get(`${this.geocoderBaseUrl}/suggest?`, { params: params })
            .pipe(map((suggestResultObject) => {
            const collations = this.formatCollations(suggestResultObject.spellcheck.collations);
            const places = this.formatPlaces(suggestResultObject);
            return { collations, places };
        }));
    }
    lookup(id, options) {
        let params = {
            id: id,
            fl: '*'
        };
        if (options) {
            params = Object.assign(params, options);
        }
        return this.http.get(`${this.geocoderBaseUrl}/lookup?`, { params: params }).toPromise().then((lookupResponse) => {
            return this.formatLookupResponse(lookupResponse);
        });
    }
    lookup$(id, options) {
        let params = {
            id: id,
            fl: '*'
        };
        if (options) {
            params = Object.assign(params, options);
        }
        return this.http.get(`${this.geocoderBaseUrl}/lookup?`, { params: params }).pipe(map((lookupResponse) => this.formatLookupResponse(lookupResponse)));
    }
    free(searchTerm, options) {
        let params = {
            q: searchTerm,
            fl: '*',
            fq: '*',
            rows: 10,
            start: 0,
        };
        if (options) {
            params = Object.assign(params, options);
        }
        return this.http.get(`${this.geocoderBaseUrl}/free?`, { params: params })
            .toPromise().then((freeResponse) => {
            return this.formatReverseResponse(freeResponse);
        });
    }
    free$(searchTerm, options) {
        let params = {
            q: searchTerm,
            fl: '*',
            fq: '*',
            rows: 10,
            start: 0,
        };
        if (options) {
            params = Object.assign(params, options);
        }
        return this.http.get(`${this.geocoderBaseUrl}/free?`, { params: params }).pipe(map((freeResponse) => this.formatReverseResponse(freeResponse)));
    }
    reverse(location, options) {
        let params = {
            type: 'adres',
            fq: '*',
            fl: '*',
            rows: 10,
            distance: 200 // meter,
        };
        if (options) {
            params = Object.assign(params, options, location);
        }
        const reverseUrl = 'https://geodata.nationaalgeoregister.nl/locatieserver/revgeo?';
        return this.http.get(reverseUrl, { params: params }).toPromise().then((reverseResponse) => {
            return this.formatReverseResponse(reverseResponse);
        });
    }
    reverse$(location, options) {
        let params = {
            type: 'adres',
            fq: '*',
            fl: '*',
            rows: 10,
            distance: 200 // meter,
        };
        if (options) {
            params = Object.assign(params, options, location);
        }
        const reverseUrl = 'https://geodata.nationaalgeoregister.nl/locatieserver/revgeo?';
        return this.http.get(reverseUrl, { params: params }).pipe(map((reverseResponse) => this.formatReverseResponse(reverseResponse)));
    }
    formatCollations(collations) {
        const parsedCollations = [];
        for (let i = 0; i < collations.length; i += 2) {
            const collation = {
                id: i,
                naam: collations[i + 1].misspellingsAndCorrections[1],
                weergavenaam: `${collations[i + 1].hits} resultaten gevonden voor <strong> ${collations[i + 1].misspellingsAndCorrections[1]} </strong>`,
                hits: collations[i + 1].hits,
            };
            parsedCollations.push(collation);
        }
        return parsedCollations;
    }
    formatPlaces(suggestResponse) {
        const places = suggestResponse.response.docs.map((place) => {
            return Object.assign(Object.assign({}, place), { highlight: suggestResponse.highlighting[place.id].suggest[0] });
        });
        return places;
    }
    /**
     * Parse WKT in lookup response.
     */
    formatLookupResponse(lookupResponse) {
        const formatted = lookupResponse.response.docs.map((lookupResult) => {
            const formattedLookupResult = lookupResult;
            if (lookupResult.centroide_ll) {
                formattedLookupResult.centroide_ll = parse(lookupResult.centroide_ll);
            }
            if (lookupResult.centroide_rd) {
                formattedLookupResult.centroide_rd = parse(lookupResult.centroide_rd);
            }
            if (lookupResult.geometrie_rd) {
                formattedLookupResult.geometrie_rd = parse(lookupResult.geometrie_rd);
            }
            if (lookupResult.geometrie_ll) {
                formattedLookupResult.geometrie_ll = parse(lookupResult.geometrie_ll);
            }
            return formattedLookupResult;
        });
        return formatted[0] || {};
    }
    formatReverseResponse(lookupResultObject) {
        const formatted = lookupResultObject.response.docs.map((reverseResult) => {
            const formattedLookupResult = reverseResult;
            if (reverseResult.centroide_ll) {
                formattedLookupResult.centroide_ll = parse(reverseResult.centroide_ll);
            }
            if (reverseResult.centroide_rd) {
                formattedLookupResult.centroide_rd = parse(reverseResult.centroide_rd);
            }
            if (reverseResult.geometrie_rd) {
                formattedLookupResult.geometrie_rd = parse(reverseResult.geometrie_rd);
            }
            if (reverseResult.geometrie_ll) {
                formattedLookupResult.geometrie_ll = parse(reverseResult.geometrie_ll);
            }
            formattedLookupResult.highlight = reverseResult.weergavenaam;
            return formattedLookupResult;
        });
        return formatted;
    }
}
GeocoderService.ɵfac = function GeocoderService_Factory(t) { return new (t || GeocoderService)(ɵngcc0.ɵɵinject(ɵngcc1.HttpClient)); };
GeocoderService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: GeocoderService, factory: GeocoderService.ɵfac });
GeocoderService.ctorParameters = () => [
    { type: HttpClient }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GeocoderService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.HttpClient }]; }, null); })();

// Leaflet requires the old school way for importing
class GeocoderComponent {
    constructor(geocoderService) {
        this.geocoderService = geocoderService;
        this.resultCountLimit = 10;
        this.resultFieldsToShow = [];
        this.placeFound = new EventEmitter();
        this.formattedPlaceholder = '';
        this.resultsVisible = true;
        this.places = [];
        this.collations = [];
        this.searchThreshold = 2;
        this.foundPlace = null;
        this.selectedItem = [];
        this.selectedIndex = -1;
        this.subscriptions = [];
        this.searchField = new FormControl();
        this.searchForm = new FormGroup({ search: this.searchField });
        this.subscriptions.push(this.subscribeToSearchFieldValueChanges());
    }
    clickout(event) {
        if (this.geocoderRef.nativeElement.contains(event.target)) {
            this.showResults();
        }
        else {
            this.hideResults();
        }
    }
    ngOnInit() {
        this.formattedPlaceholder = this.formatPlaceholder(this.placeholder);
    }
    ngOnChanges(changes) {
        if (changes.searchInput) {
            this.searchField.setValue(changes.searchInput.currentValue);
        }
    }
    ngOnDestroy() {
        this.subscriptions.forEach(s => s.unsubscribe());
    }
    subscribeToSearchFieldValueChanges() {
        return this.searchField.valueChanges.pipe(filter(searchInput => searchInput.length > this.searchThreshold), debounceTime(100), distinctUntilChanged(), switchMap(searchInput => this.suggest(searchInput))).subscribe(suggestResponse => {
            this.places = suggestResponse.places;
            this.collations = suggestResponse.collations;
        }); // Actual call is fired.
    }
    suggest(searchTerm) {
        const flArray = ['id', 'score', 'type', 'weergavenaam', ...this.resultFieldsToShow];
        const params = {
            fq: '*:*',
            fl: flArray.toString(),
            rows: this.resultCountLimit
        };
        if (this.type) {
            const replace = ',';
            const regex = new RegExp(replace, 'g');
            const query = this.type.replace(regex, ' OR ');
            params.fq = `type:(${query})`;
            if (this.province) {
                params.fq += ` AND provincienaam:(${this.province})`;
            }
        }
        else if (this.province) {
            params.fq = `provincienaam:(${this.province})`;
        }
        return this.geocoderService.suggest$(searchTerm, params);
    }
    free() {
        const searchTerm = this.searchField.value;
        const params = { fq: '*:*' };
        if (this.type) {
            const replace = ',';
            const regex = new RegExp(replace, 'g');
            const query = this.type.replace(regex, ' OR ');
            params.fq = `type:(${query})`;
            if (this.province) {
                params.fq += ` AND provincienaam:(${this.province})`;
            }
        }
        else if (this.province) {
            params.fq = `provincienaam:(${this.province})`;
        }
        return this.geocoderService.free$(searchTerm, params).subscribe(places => {
            this.places = places;
        });
    }
    lookup(id) {
        this.geocoderService.lookup$(id).subscribe((lookupObj) => {
            this.foundPlace = lookupObj;
            if (this.resultFieldsToShow.length > 0) {
                const input = this.resultFieldsToShow.map((resultFieldToShow) => {
                    return this.foundPlace[resultFieldToShow];
                }).join(', ');
                this.fillInput(input);
            }
            else {
                this.fillInput(this.foundPlace.weergavenaam);
            }
            this.placeFound.emit(this.foundPlace);
            this.clearPlaces();
        });
    }
    clear() {
        this.foundPlace = null;
        this.searchField.setValue('');
        this.clearPlaces();
    }
    clearPlaces() {
        this.resetIndex();
        this.places = [];
        this.collations = [];
    }
    handleEnter() {
        if (this.selectedIndex === -1) {
            return;
        }
        const selectedPlace = this.places[this.selectedIndex];
        this.lookup(selectedPlace.id);
    }
    isNoResultsFound() {
        const reachedThreshold = (this.searchField.value && this.searchField.value.length) > this.searchThreshold;
        const noSuggestions = this.places.length === 0;
        const noResult = (this.foundPlace == null);
        return (reachedThreshold) && (noSuggestions) && (noResult);
    }
    showCollations() {
        return this.collations.length > 0 && this.places.length === 0;
    }
    fillInput(content, emitEvent = false) {
        this.searchField.setValue(content, {
            emitEvent: emitEvent
        });
    }
    isHighlighted(i) {
        if (i === this.selectedIndex) {
            return true;
        }
    }
    moveUp() {
        if (this.selectedIndex > 0) {
            this.selectedIndex--;
        }
    }
    moveDown() {
        if (this.selectedIndex < this.places.length) {
            this.selectedIndex++;
        }
    }
    resetIndex() {
        this.selectedIndex = -1;
    }
    canQuery() {
        const searchInput = this.searchField.value;
        return searchInput && searchInput.length > this.searchThreshold;
    }
    canClear() {
        const searchInput = this.searchField.value;
        return searchInput && searchInput.length > 0;
    }
    formatPlaceholder(placeholder) {
        let placeholderText = 'Zoeken op de kaart...';
        if (placeholder) {
            placeholderText = placeholder;
        }
        return placeholderText;
    }
    onCollationClick(name) {
        this.fillInput(name, true);
    }
    showResults() {
        this.resultsVisible = true;
    }
    hideResults() {
        this.resultsVisible = false;
    }
}
GeocoderComponent.ɵfac = function GeocoderComponent_Factory(t) { return new (t || GeocoderComponent)(ɵngcc0.ɵɵdirectiveInject(GeocoderService)); };
GeocoderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: GeocoderComponent, selectors: [["geocoder"]], viewQuery: function GeocoderComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.geocoderRef = _t.first);
    } }, hostBindings: function GeocoderComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function GeocoderComponent_click_HostBindingHandler($event) { return ctx.clickout($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { resultCountLimit: "resultCountLimit", resultFieldsToShow: "resultFieldsToShow", type: "type", searchInput: "searchInput", placeholder: "placeholder", province: "province" }, outputs: { placeFound: "placeFound" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 12, vars: 9, consts: [[1, "geocoder-container", 3, "keydown.ArrowUp", "keydown.ArrowDown", "keydown.Enter"], ["geocoder", ""], [1, "geocoder-search-container"], [3, "formGroup"], ["formControlName", "search", "type", "text", "aria-label", "Zoeken", "maxlength", "100", 1, "geocoder-search-input", 3, "placeholder", "focus"], ["aria-label", "zoeken", 1, "searchButton", 3, "ngClass", "click"], ["aria-hidden", "true", "data-fa-processed", "", "data-prefix", "fal", "data-icon", "search", "role", "img", "xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 512 512", 1, "svg-inline--fa", "fa-search", "fa-w-16", "fa-9x"], ["fill", "currentColor", "d", "M508.5 481.6l-129-129c-2.3-2.3-5.3-3.5-8.5-3.5h-10.3C395 312 416 262.5 416 208 416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c54.5 0 104-21 141.1-55.2V371c0 3.2 1.3 6.2 3.5 8.5l129 129c4.7 4.7 12.3 4.7 17 0l9.9-9.9c4.7-4.7 4.7-12.3 0-17zM208 384c-97.3 0-176-78.7-176-176S110.7 32 208 32s176 78.7 176 176-78.7 176-176 176z", 1, ""], ["aria-label", "zoekopdracht verwijderen", 1, "closeButton", 3, "ngClass", "click"], ["aria-hidden", "true", "data-fa-processed", "", "data-prefix", "fal", "data-icon", "times", "role", "img", "xmlns", "http://www.w3.org/2000/svg", "viewBox", "0 0 384 512", 1, "svg-inline--fa", "fa-times", "fa-w-12", "fa-7x"], ["fill", "currentColor", "d", "M217.5 256l137.2-137.2c4.7-4.7 4.7-12.3 0-17l-8.5-8.5c-4.7-4.7-12.3-4.7-17 0L192 230.5 54.8 93.4c-4.7-4.7-12.3-4.7-17 0l-8.5 8.5c-4.7 4.7-4.7 12.3 0 17L166.5 256 29.4 393.2c-4.7 4.7-4.7 12.3 0 17l8.5 8.5c4.7 4.7 12.3 4.7 17 0L192 281.5l137.2 137.2c4.7 4.7 12.3 4.7 17 0l8.5-8.5c4.7-4.7 4.7-12.3 0-17L217.5 256z", 1, ""], ["class", "geocoder-suggestions-container", 4, "ngIf"], [1, "geocoder-suggestions-container"], ["class", "geocoder-suggestions-results", 4, "ngIf"], ["class", "geocoder-no-results", 4, "ngIf"], ["class", "geocoder-collation-container", 4, "ngIf"], [1, "geocoder-suggestions-results"], [1, "geocoder-suggestions-list"], ["class", "geocoder-suggestion", 3, "tabindex", "ngClass", "click", 4, "ngFor", "ngForOf"], [1, "geocoder-suggestion", 3, "tabindex", "ngClass", "click"], ["class", "geocoder-suggestion-title", 3, "innerHTML", 4, "ngIf"], ["class", "geocoder-suggestion-title", 4, "ngIf"], [1, "geocoder-suggestion-type"], [1, "geocoder-suggestion-title", 3, "innerHTML"], [1, "geocoder-suggestion-title"], [4, "ngFor", "ngForOf"], [1, "geocoder-no-results"], [1, "geocoder-collation-container"], [1, "geocoder-collations-list"], ["class", "geocoder-collation", 3, "tabindex", "innerHTML", "click", 4, "ngFor", "ngForOf"], [1, "geocoder-collation", 3, "tabindex", "innerHTML", "click"]], template: function GeocoderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵlistener("keydown.ArrowUp", function GeocoderComponent_Template_div_keydown_ArrowUp_0_listener() { return ctx.moveUp(); })("keydown.ArrowDown", function GeocoderComponent_Template_div_keydown_ArrowDown_0_listener() { return ctx.moveDown(); })("keydown.Enter", function GeocoderComponent_Template_div_keydown_Enter_0_listener() { return ctx.handleEnter(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "form", 3);
        ɵngcc0.ɵɵelementStart(4, "input", 4);
        ɵngcc0.ɵɵlistener("focus", function GeocoderComponent_Template_input_focus_4_listener() { return ctx.showResults(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "button", 5);
        ɵngcc0.ɵɵlistener("click", function GeocoderComponent_Template_button_click_5_listener() { return ctx.free(); });
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(6, "svg", 6);
        ɵngcc0.ɵɵelement(7, "path", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵnamespaceHTML();
        ɵngcc0.ɵɵelementStart(8, "button", 8);
        ɵngcc0.ɵɵlistener("click", function GeocoderComponent_Template_button_click_8_listener() { return ctx.clear(); });
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(9, "svg", 9);
        ɵngcc0.ɵɵelement(10, "path", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, GeocoderComponent_div_11_Template, 4, 3, "div", 11);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("formGroup", ctx.searchForm);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("placeholder", ctx.formattedPlaceholder);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c2, ctx.canQuery()));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(7, _c2, ctx.canClear()));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.resultsVisible === true);
    } }, directives: [ɵngcc2.ɵangular_packages_forms_forms_y, ɵngcc2.NgControlStatusGroup, ɵngcc2.FormGroupDirective, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc2.MaxLengthValidator, ɵngcc3.NgClass, ɵngcc3.NgIf, ɵngcc3.NgForOf], styles: ["[_ngcontent-%COMP%]::-webkit-input-placeholder{color:#797979;font-size:14px}[_ngcontent-%COMP%]:-moz-placeholder, [_ngcontent-%COMP%]::-moz-placeholder{color:#797979;font-size:14px}[_ngcontent-%COMP%]:-ms-input-placeholder{color:#797979;font-size:14px}[_ngcontent-%COMP%]::-ms-input-placeholder{color:#797979;font-size:14px}[_ngcontent-%COMP%]:-moz-placeholder-shown{color:#797979;font-size:14px}[_ngcontent-%COMP%]:placeholder-shown{color:#797979;font-size:14px}[_ngcontent-%COMP%]:focus{outline:0}input[_ngcontent-%COMP%], li[_ngcontent-%COMP%], p[_ngcontent-%COMP%], span[_ngcontent-%COMP%]{color:#797979;font-size:14px;margin:0}.geocoder-container[_ngcontent-%COMP%]{max-width:100%;position:relative;width:100%}.geocoder-search-container[_ngcontent-%COMP%]{background:#fff;border:none;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);width:100%}.geocoder-search-input[_ngcontent-%COMP%]{border:none;max-width:100%;overflow:hidden;padding:15px;width:95%;width:calc(100% - 80px)}button[_ngcontent-%COMP%]{cursor:pointer;position:absolute}.searchButton[_ngcontent-%COMP%]{background:#fff;border:none;border-right:1px solid #797979;padding-right:15px;right:50px;top:13px}.closeButton[_ngcontent-%COMP%]{background:#fff;border:none;margin-left:16px;right:10px;top:10px}svg[_ngcontent-%COMP%]{color:#797979;width:18px}.closeButton[_ngcontent-%COMP%]:hover   svg[_ngcontent-%COMP%], .highlight[_ngcontent-%COMP%]   svg[_ngcontent-%COMP%], .searchButton[_ngcontent-%COMP%]:hover   svg[_ngcontent-%COMP%]{color:#79abff;transition:all .1s cubic-bezier(.075,.82,.165,1)}.geocoder-suggestions-results[_ngcontent-%COMP%]{background:#fff;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);max-height:400px;overflow-y:auto;position:absolute;top:53px;width:100%;z-index:1}.geocoder-collations-list[_ngcontent-%COMP%], .geocoder-suggestions-list[_ngcontent-%COMP%]{list-style-type:none;margin:0;padding:0}.geocoder-suggestion[_ngcontent-%COMP%]{border-bottom:1px solid #f1f1f1;padding:15px;position:relative}.geocoder-suggestion-title[_ngcontent-%COMP%]{max-width:243px}.geocoder-suggestion-type[_ngcontent-%COMP%]{color:#bfbfbf;font-size:11px;font-weight:700;position:absolute;right:15px;text-transform:uppercase;top:36%}.geocoder-collation[_ngcontent-%COMP%]:hover, .geocoder-suggestion[_ngcontent-%COMP%]:hover{background-color:#f7f7f7;cursor:pointer}.focus-background[_ngcontent-%COMP%]{background-color:#f7f7f7}.geocoder-no-results[_ngcontent-%COMP%]{background:#fff;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);box-sizing:border-box;padding:15px;position:absolute;top:53px;width:100%;z-index:1}.geocoder-collation[_ngcontent-%COMP%]{z-index:1}.geocoder-collation-container[_ngcontent-%COMP%]{background:#fff;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);position:absolute;top:108px;width:100%;z-index:1}.geocoder-collation-container[_ngcontent-%COMP%]   li[_ngcontent-%COMP%]{border-bottom:1px solid #f1f1f1;padding:15px}.searchButton[_ngcontent-%COMP%]{height:20px}.closeButton[_ngcontent-%COMP%]{height:26px}"] });
GeocoderComponent.ctorParameters = () => [
    { type: GeocoderService }
];
GeocoderComponent.propDecorators = {
    geocoderRef: [{ type: ViewChild, args: ['geocoder', { static: false },] }],
    type: [{ type: Input }],
    searchInput: [{ type: Input }],
    placeholder: [{ type: Input }],
    resultCountLimit: [{ type: Input }],
    resultFieldsToShow: [{ type: Input }],
    province: [{ type: Input }],
    placeFound: [{ type: Output }],
    clickout: [{ type: HostListener, args: ['document:click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GeocoderComponent, [{
        type: Component,
        args: [{
                template: "<div #geocoder (keydown.ArrowUp)=\"moveUp()\" (keydown.ArrowDown)=\"moveDown()\" (keydown.Enter)=\"handleEnter()\"\r\n  class=\"geocoder-container\">\r\n  <!-- Input -->\r\n  <div class=\"geocoder-search-container\">\r\n    <form [formGroup]=\"searchForm\">\r\n      <input (focus)=\"showResults()\" formControlName=\"search\" class=\"geocoder-search-input\" [placeholder]=\"formattedPlaceholder\" type=\"text\"\r\n        aria-label=\"Zoeken\" maxlength=\"100\">\r\n    </form>\r\n    <button class=\"searchButton\" aria-label=\"zoeken\" (click)=\"free()\" [ngClass]=\"{'highlight': canQuery()}\"> <svg\r\n        aria-hidden=\"true\" data-fa-processed=\"\" data-prefix=\"fal\" data-icon=\"search\" role=\"img\"\r\n        xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 512 512\" class=\"svg-inline--fa fa-search fa-w-16 fa-9x\">\r\n        <path fill=\"currentColor\"\r\n          d=\"M508.5 481.6l-129-129c-2.3-2.3-5.3-3.5-8.5-3.5h-10.3C395 312 416 262.5 416 208 416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c54.5 0 104-21 141.1-55.2V371c0 3.2 1.3 6.2 3.5 8.5l129 129c4.7 4.7 12.3 4.7 17 0l9.9-9.9c4.7-4.7 4.7-12.3 0-17zM208 384c-97.3 0-176-78.7-176-176S110.7 32 208 32s176 78.7 176 176-78.7 176-176 176z\"\r\n          class=\"\"></path>\r\n      </svg></button>\r\n    <button class=\"closeButton\" aria-label=\"zoekopdracht verwijderen\" (click)=\"clear()\"\r\n      [ngClass]=\"{'highlight': canClear()}\"> <svg aria-hidden=\"true\" data-fa-processed=\"\" data-prefix=\"fal\"\r\n        data-icon=\"times\" role=\"img\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 384 512\"\r\n        class=\"svg-inline--fa fa-times fa-w-12 fa-7x\">\r\n        <path fill=\"currentColor\"\r\n          d=\"M217.5 256l137.2-137.2c4.7-4.7 4.7-12.3 0-17l-8.5-8.5c-4.7-4.7-12.3-4.7-17 0L192 230.5 54.8 93.4c-4.7-4.7-12.3-4.7-17 0l-8.5 8.5c-4.7 4.7-4.7 12.3 0 17L166.5 256 29.4 393.2c-4.7 4.7-4.7 12.3 0 17l8.5 8.5c4.7 4.7 12.3 4.7 17 0L192 281.5l137.2 137.2c4.7 4.7 12.3 4.7 17 0l8.5-8.5c4.7-4.7 4.7-12.3 0-17L217.5 256z\"\r\n          class=\"\"></path>\r\n      </svg></button>\r\n  </div>\r\n  <!-- Search results-->\r\n  <div class=\"geocoder-suggestions-container\" *ngIf=\"resultsVisible === true\">\r\n    <div class=\"geocoder-suggestions-results\" *ngIf=\"places.length > 0\">\r\n      <ul class=\"geocoder-suggestions-list\">\r\n        <li class=\"geocoder-suggestion\" tabindex={{i}} *ngFor=\"let place of places; let i = index;\"\r\n          (click)=\"lookup(place.id)\" [ngClass]=\"{'focus-background': isHighlighted(i)}\">\r\n          <p *ngIf=\"resultFieldsToShow.length === 0\" class=\"geocoder-suggestion-title\" [innerHTML]=\"place.highlight\">\r\n          </p>\r\n          <p *ngIf=\"resultFieldsToShow.length > 0\" class=\"geocoder-suggestion-title\">\r\n            <ng-container *ngFor=\"let field of resultFieldsToShow\">\r\n              {{place[field]}}\r\n            </ng-container>\r\n          </p>\r\n          <p class=\"geocoder-suggestion-type\"> {{place.type}} </p>\r\n        </li>\r\n      </ul>\r\n    </div>\r\n    <div class=\"geocoder-no-results\" *ngIf=\"isNoResultsFound()\">\r\n      <p>Geen zoekresultaten gevonden voor <strong> {{searchField.value}}</strong>.</p>\r\n    </div>\r\n    <div class=\"geocoder-collation-container\" *ngIf=\"showCollations()\">\r\n      <ul class=\"geocoder-collations-list\">\r\n        <li class=\"geocoder-collation\" tabindex={{i}} (click)=\"onCollationClick(collation.naam)\"\r\n          *ngFor=\"let collation of collations; let i = index;\" [innerHTML]=\"collation.weergavenaam\">\r\n        </li>\r\n      </ul>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                selector: 'geocoder',
                styles: ["::-webkit-input-placeholder{color:#797979;font-size:14px}:-moz-placeholder,::-moz-placeholder{color:#797979;font-size:14px}:-ms-input-placeholder{color:#797979;font-size:14px}::-ms-input-placeholder{color:#797979;font-size:14px}:-moz-placeholder-shown{color:#797979;font-size:14px}:placeholder-shown{color:#797979;font-size:14px}:focus{outline:0}input,li,p,span{color:#797979;font-size:14px;margin:0}.geocoder-container{max-width:100%;position:relative;width:100%}.geocoder-search-container{background:#fff;border:none;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);width:100%}.geocoder-search-input{border:none;max-width:100%;overflow:hidden;padding:15px;width:95%;width:calc(100% - 80px)}button{cursor:pointer;position:absolute}.searchButton{background:#fff;border:none;border-right:1px solid #797979;padding-right:15px;right:50px;top:13px}.closeButton{background:#fff;border:none;margin-left:16px;right:10px;top:10px}svg{color:#797979;width:18px}.closeButton:hover svg,.highlight svg,.searchButton:hover svg{color:#79abff;transition:all .1s cubic-bezier(.075,.82,.165,1)}.geocoder-suggestions-results{background:#fff;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);max-height:400px;overflow-y:auto;position:absolute;top:53px;width:100%;z-index:1}.geocoder-collations-list,.geocoder-suggestions-list{list-style-type:none;margin:0;padding:0}.geocoder-suggestion{border-bottom:1px solid #f1f1f1;padding:15px;position:relative}.geocoder-suggestion-title{max-width:243px}.geocoder-suggestion-type{color:#bfbfbf;font-size:11px;font-weight:700;position:absolute;right:15px;text-transform:uppercase;top:36%}.geocoder-collation:hover,.geocoder-suggestion:hover{background-color:#f7f7f7;cursor:pointer}.focus-background{background-color:#f7f7f7}.geocoder-no-results{background:#fff;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);box-sizing:border-box;padding:15px;position:absolute;top:53px;width:100%;z-index:1}.geocoder-collation{z-index:1}.geocoder-collation-container{background:#fff;box-shadow:0 3px 6px rgba(0,0,0,.16),0 3px 6px rgba(0,0,0,.23);position:absolute;top:108px;width:100%;z-index:1}.geocoder-collation-container li{border-bottom:1px solid #f1f1f1;padding:15px}.searchButton{height:20px}.closeButton{height:26px}"]
            }]
    }], function () { return [{ type: GeocoderService }]; }, { resultCountLimit: [{
            type: Input
        }], resultFieldsToShow: [{
            type: Input
        }], placeFound: [{
            type: Output
        }], clickout: [{
            type: HostListener,
            args: ['document:click', ['$event']]
        }], geocoderRef: [{
            type: ViewChild,
            args: ['geocoder', { static: false }]
        }], type: [{
            type: Input
        }], searchInput: [{
            type: Input
        }], placeholder: [{
            type: Input
        }], province: [{
            type: Input
        }] }); })();

class GeocoderModule {
}
GeocoderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: GeocoderModule });
GeocoderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function GeocoderModule_Factory(t) { return new (t || GeocoderModule)(); }, providers: [GeocoderService], imports: [[
            HttpClientModule,
            CommonModule,
            ReactiveFormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(GeocoderModule, { declarations: function () { return [GeocoderComponent]; }, imports: function () { return [HttpClientModule,
        CommonModule,
        ReactiveFormsModule]; }, exports: function () { return [GeocoderComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GeocoderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    HttpClientModule,
                    CommonModule,
                    ReactiveFormsModule
                ],
                exports: [GeocoderComponent],
                declarations: [GeocoderComponent],
                providers: [GeocoderService]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { GeocoderModule, GeocoderService, GeocoderComponent as ɵa };
