// angular-geocoder/esm2015/app/geocoder/locatieserver.model.js
// https://github.com/PDOK/locatieserver/wiki/API-Locatieserver
export class BasicGeocoderOptions {
}
export var ResponseTypes;
(function (ResponseTypes) {
    ResponseTypes[ResponseTypes["provincie"] = 0] = "provincie";
    ResponseTypes[ResponseTypes["gemeente"] = 1] = "gemeente";
    ResponseTypes[ResponseTypes["woonplaats"] = 2] = "woonplaats";
    ResponseTypes[ResponseTypes["weg"] = 3] = "weg";
    ResponseTypes[ResponseTypes["postcode"] = 4] = "postcode";
    ResponseTypes[ResponseTypes["adres"] = 5] = "adres";
    ResponseTypes[ResponseTypes["perceel"] = 6] = "perceel";
    ResponseTypes[ResponseTypes["hectometerpaal"] = 7] = "hectometerpaal";
    ResponseTypes[ResponseTypes["wijk"] = 8] = "wijk";
    ResponseTypes[ResponseTypes["buurt"] = 9] = "buurt";
    ResponseTypes[ResponseTypes["waterschapsgrens"] = 10] = "waterschapsgrens";
    ResponseTypes[ResponseTypes["appartementsrecht"] = 11] = "appartementsrecht";
})(ResponseTypes || (ResponseTypes = {}));
export var ResponseSources;
(function (ResponseSources) {
    ResponseSources[ResponseSources["BAG"] = 0] = "BAG";
    ResponseSources[ResponseSources["NWB"] = 1] = "NWB";
    ResponseSources[ResponseSources["BAG/NWB"] = 2] = "BAG/NWB";
    ResponseSources[ResponseSources["DKK"] = 3] = "DKK";
    ResponseSources[ResponseSources["Bestuurlijke Grenzen"] = 4] = "Bestuurlijke Grenzen";
    ResponseSources[ResponseSources["CBS"] = 5] = "CBS";
    ResponseSources[ResponseSources["HWH"] = 6] = "HWH"; // Waterschapshuis
})(ResponseSources || (ResponseSources = {}));
export class SuggestOptions extends BasicGeocoderOptions {
}
export class LookupOptions extends BasicGeocoderOptions {
}
export class ReverseOptions extends BasicGeocoderOptions {
}
export class SuggestResponse {
}
export class SuggestObject {
}
export class LookupResponse {
}
export class LookupObject {
}
export class ReverseResponse {
}
export class ReverseGeometry {
}
